### ISSUE TITLE
trimmed/truncated mean nomenclature

### ISSUE THREAD

**Issue by moeller0** [network-quality/draft-ietf-ippm-responsiveness#78]
The current draft states:
"Over the timeframe of these intervals a potentially large number of samples has been collected. These may be affected by noise in the measurements, and outliers. Thus, to aggregate these we suggest to use a trimmed mean at the TMP (Trimmed Mean Percentage - default to 95%) percentile, thus providing the following numbers: TM(tcp_f), TM(tls_f), TM(http_f), TM(http_l)."

According to e.g. wikipedia:
https://en.wikipedia.org/wiki/Truncated_mean
"A truncated mean or trimmed mean is a [statistical](https://en.wikipedia.org/wiki/Statistics) [measure of central tendency](https://en.wikipedia.org/wiki/Average), much like the [mean](https://en.wikipedia.org/wiki/Mean) and [median](https://en.wikipedia.org/wiki/Median). It involves the calculation of the mean after discarding given parts of a [probability distribution](https://en.wikipedia.org/wiki/Probability_distribution) or [sample](https://en.wikipedia.org/wiki/Sampling_(statistics)) at the high and low end, and typically discarding an equal amount of both. This number of points to be discarded is usually given as a percentage of the total number of points, but may also be given as a fixed number of points."

By that definition a TPM of 95% seems to imply that 95% of the samples are excluded which I would assume is not what is intended. I would propose to change the draft text to be explicit and report the upper and lower quantiles for the included samples instead e.g.:
"(Trimmed Mean Percentage - default to including the central 95% of samples, from the 2.5% to the 97.5% quantile)"
or similar.

**Comment by cpaasch** [comment]
Thanks, that's a good point! We need to better spell out which part of the data-set is trimmed. As this is a latency-measurement it is usually sufficient to only trim at the top end, not the bottom. So, 95% trimmed mean, means that the top 5% of the data-values should be removed.

**Comment by moeller0** [comment]
Mmmh, cutting a distribution with a positive skew (like RTT/OWD data) only at the right side will end up reducing the numeric value of the average... 
Having looked mostly at CDFs in the last moths I would intuitively simply pick a quantile for each individual measure instead of an average (averaged and standard deviation are great for symmetric near-normal distributions, RTT/OWD however are neither, so I think there should better single value descriptors available).
Here is another question, why is the proposal to treat the 4 measures independently and calculate a truncated mean individually for each component instead of calculating the aggregate measures for all X repetitions and then only take the truncated mean from this distribution instead? Or at least for those "numbers" that exist in equal quantity?

**Comment by cpaasch** [comment]
> Mmmh, cutting a distribution with a positive skew (like RTT/OWD data) only at the right side will end up reducing the numeric value of the average...

Yes! Is it a problem that the overall numeric value of the average ends up being reduced slightly ?

> Having looked mostly at CDFs in the last moths I would intuitively simply pick a quantile for each individual measure instead of an average (averaged and standard deviation are great for symmetric near-normal distributions, RTT/OWD however are neither, so I think there should better single value descriptors available).

"pick a quantile" ? You mean, reporting for example 75th percentile? The problem with that is then that this creates an incentive to optimize for the percentile and not the overall distribution. I recommend https://www.youtube.com/watch?v=_uaaCiyJCFA - it's a great talk! Trimmed Mean is not perfect but among all the single-number values it seems to be the one with the least drawbacks.

> Here is another question, why is the proposal to treat the 4 measures independently and calculate a truncated mean individually for each component instead of calculating the aggregate measures for all X repetitions and then only take the truncated mean from this distribution instead? Or at least for those "numbers" that exist in equal quantity?

I'm not sure I understand 100% what you are suggesting, but are you saying that all sets of numbers tcp_f, tls_f, http_f, http_s should be used in a single data-set and we take the trimmed mean of that?

There is a big problem with that approach because these values experience significantly different latencies. Many networks prioritize TCP-handshake packets. So, you will see that those packets have much lower latencies than http_f and http_s.

**Comment by moeller0** [comment]
> Yes, problem is that results start to have extreme variance between the runs, which makes the measurement less and less useful.

This is a sign that the number of samples is too low... now, I understand that RPM needs a reasonable short run time, after all users are conditioned on 10-20 seconds for a capacity test, so dealing with high variance via increasing the N is problematic for RPM's goals.




> Did you try p99 or p95 ? We did and the variance between test-runs makes the results basically impossible to work with. (same comment as above)

Yes, I wrote silly code to extract a set of high percentiles for cake-autorate delay measurements, but so far I only looked at low load samples collected over multiple minutes (typically 10) so not all that representative for RPM's use-case; however the numbers are pretty repeatable. Also this is from simplistic ICMP delay probes, nothing involving more elaborate processing.



> In order for the responsiveness measurement to have meaning to the end-user and being able to use it as a generally acceptable metric, it needs to provide a certain stability.

+1



> Having a P99 or TM99 both will account for the long-delay packets. The advantage is that the TM99 does not discount the lower 99% while the P99 will discount them.

But here we are fine with "discounting" the faster samples. I think of latency in a budget sense, for a given use-case X ms of delay are acceptable (e.g. because of sizing of de-jitter buffers or similar) and if the application is interactive I will not care how often the delay was considerably lower but really only how often it was above the acceptable limit. This is why I prefer a high precentile over an average (but I take you point about measurement stability and acceptable test duration as well).

> What is `tcp_f(x)` ? What is `x` ?

This is short hand for the value of tcp_f at "iteration" x. Assuming RPM measures x iterations and at each measurement time it collects all 4 variables. So tcp_f() id the array of values for tcp_f, and x is the index (sorry matlab/octave use () both for functions and arrays, and since I use matlab at work I failed to see the ambiguity of that notation).



> > > Yes! Is it a problem that the overall numeric value of the average ends up being reduced slightly ?
> 
> > That is a valid question, I guess. Personally I would account unexplained variance as data and not as noise, so would aim to err on the low side for RPM.
> 
> Yes, problem is that results start to have extreme variance between the runs, which makes the measurement less and less useful.
> 
> > > "pick a quantile" ? You mean, reporting for example 75th percentile? The problem with that is then that this creates an incentive to optimize for the percentile and not the overall distribution.
> 
> > Oh, I would pick something like P95 or even P99, making it hard to gamble... show me that ISP that manages to deliver a low P95 robustly and reliably, but then screws the average latency up royaly with the remaining 5% of samples?
> 
> Did you try p99 or p95 ? We did and the variance between test-runs makes the results basically impossible to work with. (same comment as above)
> 
> In order for the responsiveness measurement to have meaning to the end-user and being able to use it as a generally acceptable metric, it needs to provide a certain stability.
> 
> > > I recommend https://www.youtube.com/watch?v=_uaaCiyJCFA - it's a great talk! Trimmed Mean is not perfect but among all the single-number values it seems to be the one with the least drawbacks.
> 
> > I watched that talk, but while it has valid points (especially the look at the histogram), I do not think I buy the "optimize for PXX" argument, as with a sufficiently high percentile/quantile it becomes hard to screw things up, especially if one offers a truncated mean as alternative which will also tend to ignore the highest 5% of samples. I guess my point is for low latency applications it helps little if the average delay is low, if there is a sufficiently high number of long-delay packets in the mix...
> 
> Having a P99 or TM99 both will account for the long-delay packets. The advantage is that the TM99 does not discount the lower 99% while the P99 will discount them.
> 
> > > I'm not sure I understand 100% what you are suggesting, but are you saying that all sets of numbers tcp_f, tls_f, http_f, http_s should be used in a single data-set and we take the trimmed mean of that?
> 
> > Ah, sorry, no I was describing this badly, let me try again:
> 
> > Currently you have:
> > Responsiveness = 60000 /
> > (1/6*(TM(tcp_f) + TM(tls_f) + TM(http_f)) + 1/2*TM(http_s))
> 
> > and I ask why not use:
> 
> > TM( collection over all t of (1/6*(tcp_f(x) + tls_f(x) + http_f(x)) + 1/2*http_s(x)))
> 
> What is `tcp_f(x)` ? What is `x` ? Because, in the draft `tcp_f` is a set of values, not a function of a variable x, thus the notation `tcp_f(x)` is incorrect.
> 
> that is assuming that each variable was measured N times calculate the proto_responsiveness N times and the report the truncated mean over these proto_RPM values as the final RPM. However this assumes that these four measures are taken in meaningful sets, which might be a wrong assumption.
> 
> > > There is a big problem with that approach because these values experience significantly different latencies. Many networks prioritize TCP-handshake packets. So, you will see that those packets have much lower latencies than http_f and http_s.
> > > But would that be an issue for what I sketch above?
> 
> I still don't understand the above sketch.

My fault, let me try once more. Assuming we want to collect 20 iterations for each of the 4 variables and we measure them in a fixed order, e.g.:
1) tcp_f
2) tls_f
3) http_f
4) http_s

Now we have 4 arrays, one for each variable with 20 values each. The current draft proposes to "condense" each array to a single value by calculating the truncated mean and then feed these 4 condensed variables into :

Responsiveness = 60000 / (1/6*(TM(tcp_f) + TM(tls_f) + TM(http_f)) + 1/2*TM(http_s))

My idea is to calculate 20an array of 20 individual responsiveness "samples":

Responsiveness[1:20] = 60000 / (1/6*(tcp_f + tls_f + http_f) + 1/2*http_s)

And then calculate the truncated mean over this 20-value Responsiveness array.

**Comment by cpaasch** [comment]
> My fault, let me try once more. Assuming we want to collect 20 iterations for each of the 4 variables and we measure them in a fixed order, e.g.:

> tcp_f
> tls_f
> http_f
> http_s
> Now we have 4 arrays, one for each variable with 20 values each. The current draft proposes to "condense" each array to a single value by calculating the truncated mean and then feed these 4 condensed variables into :

> Responsiveness = 60000 / (1/6*(TM(tcp_f) + TM(tls_f) + TM(http_f)) + 1/2*TM(http_s))

> My idea is to calculate 20an array of 20 individual responsiveness "samples":

> Responsiveness[1:20] = 60000 / (1/6*(tcp_f + tls_f + http_f) + 1/2*http_s)

> And then calculate the truncated mean over this 20-value Responsiveness array.

I see - so it is just a different way of aggregating the values. Can you explain what you believe is the difference between one and the other ? What is the rationale behind your proposal ? What are the advantages/disadvantages ?

I can understand why treating `tcp_f`, `tls_f` and `http_f` as part of the same "set", but I don't see the correlation to `http_s`.

**Comment by cpaasch** [comment]
As discussed in the meeting. Decided to stick with the current approach.
